### PR TITLE
Validation: check raw wide pointer metadata

### LINKED ISSUE #918: Report null fat raw pointers
https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=37f152716318de72dd678cb4b8906257
As far as I know, this example is ub. I doubt 
detecting this is very important since either zeroed or transmute are probably the only way to do this, and it seems to be as insta ub as setting a reference to null (since the vtable ptr Is NonNull iirc) so i doubt any working code does it.

### PR DESCRIPTION
**Pull request by RalfJung**
While I was at it, I also added a missing check for slices not to be too big.

r? @oli-obk 
Fixes https://github.com/rust-lang/miri/issues/918

### CODE REVIEW

**Comment by oli-obk** [comment]
So https://doc.rust-lang.org/std/ptr/fn.slice_from_raw_parts.html is wrong? It's a safe function allowing the construction of such invalid slices.

It [can also be done safely on stable](https://github.com/rust-lang/rust/pull/60667#discussion_r287180043): 

```
123456789 as *const [(); usize::max_value()] as *const [()] as *const [u64]
```

Did anything change in that regard?

**Comment by RalfJung** [comment]
Ouch, I forgot about that.

What changed is that when `slice_from_raw_parts` landed, nobody was aware of https://github.com/rust-lang/rust/pull/62603.

**Comment by RalfJung** [comment]
Looks like @eddyb was the only one who even knew that we require wide raw ptr metadata to be valid.^^ Nobody in https://github.com/rust-lang/rust/pull/60667 and https://github.com/rust-lang/unsafe-code-guidelines/issues/166 knew.

**Comment by RalfJung** [comment]
For this PR, we could (for now) say the fact that this cast is allowed is a bug, and wait until https://github.com/rust-lang/rust/issues/63851 / https://github.com/rust-lang/unsafe-code-guidelines/issues/166 are resolved. That would match the UB lists in the Nomicon / Reference.

Or we do the slice length check not for raw pointers. That would make the UB documentation more messy. Or we declare the slice length to be just part of the *safety* invariant of slices so that there is nothing to check, keeping UB simpler.

File: src/librustc_mir/interpret/validity.rs [hunk 1]
```diff
@@ -250,6 +250,56 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
...
+                // FIXME: More checks for the vtable.
+            }
+            ty::Slice(..) | ty::Str => {
+                let len = try_validation!(meta.unwrap().to_usize(self.ecx),
+                    "non-integer slice length in wide pointer", self.path);
+                // check max slice length
+                let elem_size = match tail.sty {
+                    ty::Str => Size::from_bytes(1),
+                    ty::Slice(ty) => self.ecx.layout_of(ty)?.size,
+                    _ => bug!("It cannot be another type"),
```
Comment (eddyb) [inline src/librustc_mir/interpret/validity.rs]: I tend to use `unreachable!()` (or just `bug!()`, I guess) when it's statically unreachable.

I kind of wish rustc knew enough to allow this as exhaustive.

File: src/librustc_mir/interpret/validity.rs [hunk 2]
```diff
@@ -250,6 +250,56 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
...
+                };
+                if elem_size.checked_mul(len, &*self.ecx.tcx).is_none() {
```
Comment (eddyb) [inline src/librustc_mir/interpret/validity.rs]: (offtopic) @oli-obk why doesn't `TyCtxtAt` just implement the right traits to not require all of these `&*` in miri?

File: src/librustc_mir/interpret/validity.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs]: So you are saying I should remove the text? Not sure what that would improve?

But I'll be more lazy next time then. ;)

File: src/librustc_mir/interpret/validity.rs [hunk 3]
```diff
@@ -250,6 +250,56 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
...
+                    throw_validation_failure!(
+                        "too large slice (longer than isize::MAX bytes)",
+                        self.path
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs]: I will remove this check again. It is part of the "dereferencable" check, which happens inside Stacked Borrows.

**Comment by oli-obk** [comment]
`slice_from_raw_parts` is unstable, so it's not an issue. But the stable transformation is indeed problematic.

> What changed is that when slice_from_raw_parts landed, nobody was aware of #62603.

what does #62603 have to do with raw pointer metadata?

**Comment by RalfJung** [comment]
@oli-obk I removed the new slice length check.

**Comment by RalfJung** [comment]
> what does #62603 have to do with raw pointer metadata?

Sorry, copy-pasted the wrong thing. I meant https://github.com/rust-lang/rust/issues/63851.

**Comment by RalfJung** [comment]
@bors r-

I also found where the slice size check actually happens, and added a comment. @oli-obk does that seem okay? Or should I try to fix the ICE here as well?

**Comment by RalfJung** [comment]
Hm actually I cannot get it to ICE...
```rust
use std::slice;
use std::usize;

fn main() { unsafe {
    let ptr = Box::into_raw(Box::new(0u8));
    let _x = slice::from_raw_parts(ptr, usize::MAX);
} }
```

**Comment by RalfJung** [comment]
Oh wow, that's odd. When multiplying `Size` with `u64`, a check occurs, but not when multiplying `Size with `Size`?

https://github.com/rust-lang/rust/blob/7e0afdad28c4d1154176df6d35a14e738ec311af/src/librustc_target/abi/mod.rs#L341-L360

@eddyb is that deliberate?

**Comment by RalfJung** [comment]
Oh, `Size` multiplication actually is *not* doing the right kind of check for us here... it calls `checked_mul` *for `u64`*, not for `Size`.

I will remove the last commit here and make that a new PR.

**Pull request merged**